sql: Added CREATE SEQUENCE ... AS <typename> option.

mokaixu opened the pull request:
Fixes: #25110
Fixes: #29046 

Previously, unimplementedWithIssueDetail error was thrown
when an AS option in CREATE SEQUENCE ... AS was encountered.

Now we support the AS option for creating sequences which can 
take on values such as smallint/integer/bigint. Typically (i.e.
in Postgres) the AS option allows users to specify an integer
type, which dictates the min and max integer values a sequence
can take on. Before this change, the only option value we supported was
bigint/int8/int/integer, which corresponds to our default 
min and max values for sequences.

NB: In postgres, `integer` corresponds to int4. In cockroach, we use
`nakedIntType` which defaults to int8 but can be configured in the 
cluster settings.

Release note (sql change): Added CREATE SEQUENCE AS <typename> option.

mokaixu committed:
sql: Added CREATE SEQUENCE AS <typename> option.

Previously, unimplementedWithIssueDetail error was thrown
when an AS option in CREATE SEQUENCE ... AS was encountered.

Now we support the AS option for creating sequences which can
take on values such as smallint/integer/bigint. Typically (i.e.
in Postgres) the AS option allows users to specify an integer
type, which dictates the min and max integer values a sequence
can take on. Before this change, the only option value we supported was
bigint/int8/int/integer, which corresponds to our default
min and max values for sequences.

NB: In postgres, `integer` corresponds to int4. In cockroach, we use
`nakedIntType` which defaults to int8 but can be configured in the
cluster settings.

Release note (sql change): Added CREATE SEQUENCE AS <typename> option.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/57339)
<!-- Reviewable:end -->

arulajmani reviewed (changes requested):
Thanks for taking this on and putting it out so quick! :100:  Took an initial pass at this and left some preliminary thoughts, I can take a closer look once you "un-WIP" this. 

Also, I think this PR could benefit from some logictests. I would also recommend prefixing the issue numbers linked above with a `#` in your commit message -- that way, the PR gets linked to the issues and they'll get automatically closed when this merges. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57339#-:-MNdPnBdE6nt9-kZpdhK:b-g80luz)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @arulajmani, and @mokaixu)

___
*[pkg/sql/sequence.go, line 279 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57339#-MNdI1Oq80ZtIvE-h8dT:-MNdI1Oq80ZtIvE-h8dU:b-26xm5h) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9de7d968bc8a22b6e0899b66b8a3694ce508e15/pkg/sql/sequence.go#L279)):*
> ```Go
> 
> 		switch option.Name {
> 		case tree.SeqOptAs:
> ```

How would this work when explicit values for Min/Max are supplied in the sequence options? Concretely, what happens when the `MaxValue` supplied is greater than the one allowed by the datatype?
___
*[pkg/sql/parser/sql.y, line 6489 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57339#-MNdHcFZ0Ti_Td45T-3i:-MNdHcFZ0Ti_Td45T-3j:b1428jw) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9de7d968bc8a22b6e0899b66b8a3694ce508e15/pkg/sql/parser/sql.y#L6489)):*
> ```Yacc
> 
> sequence_option_elem:
>   AS typename                  { $$.val = tree.SequenceOption{Name: tree.SeqOptAs, AsTypename: $2.colType()}}
> ```

As is, this will accept all possible types, but what we really want here is to constrain this to `smalling`, `integer`, and `bigint`. As this is a subset of `numeric` types, I would recommend creating a new rule for handling this. 


<!-- Sent from Reviewable.io -->

mokaixu reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57339)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru and @arulajmani)

___
*[pkg/sql/sequence.go, line 279 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57339#-MNdI1Oq80ZtIvE-h8dT:-MNdWmX993N3gUT8loYl:b-vqp1ct) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9de7d968bc8a22b6e0899b66b8a3694ce508e15/pkg/sql/sequence.go#L279)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

How would this work when explicit values for Min/Max are supplied in the sequence options? Concretely, what happens when the `MaxValue` supplied is greater than the one allowed by the datatype?
</blockquote></details>

Would love a PM's insight on this, but I'm thinking the 2 approaches are
1) error out when the MAX VALUE set conflicts with the integer type configure
2) let the MAX VALUE override what the default max is for the integer type.

Since my change deals with int64 (which is the largest encompassing integer type), option 2) seems reasonable to me. If/when we want to implement support for int2/int4
___
*[pkg/sql/parser/sql.y, line 6489 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57339#-MNdHcFZ0Ti_Td45T-3i:-MNdWiHD3n_PiOGg0v8x:b-jg9chn) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9de7d968bc8a22b6e0899b66b8a3694ce508e15/pkg/sql/parser/sql.y#L6489)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

As is, this will accept all possible types, but what we really want here is to constrain this to `smalling`, `integer`, and `bigint`. As this is a subset of `numeric` types, I would recommend creating a new rule for handling this. 
</blockquote></details>

dare i say i'm creating ARUL?!?!??!?!?!?


<!-- Sent from Reviewable.io -->

arulajmani reviewed (changes requested):
Thanks for making changes, this is getting there. I have concerns that as is, this breaks round trip ability of `SHOW CREATE SEQUENCE`. Also, as I mentioned before, could you write some logic tests for this? I think it would greatly simplify some of the tests you have going on here. It might be worth adding some for `ALTER SEQUENCE` as well, as this patch affects both but only seems to focus on `CREATE` in its testing.

> Reviewed 5 of 14 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57339#-:-MOCQHBz3DMKAxlZqfXz:b86u0yp)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @arulajmani, and @mokaixu)

___
*[pkg/sql/sequence.go, line 279 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57339#-MNdI1Oq80ZtIvE-h8dT:-MOCRkr4Eq4pGs9B-NRk:bnuy05o) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9de7d968bc8a22b6e0899b66b8a3694ce508e15/pkg/sql/sequence.go#L279)):*
<details><summary><i>Previously, mokaixu (Monica Xu) wrote…</i></summary><blockquote>

Would love a PM's insight on this, but I'm thinking the 2 approaches are
1) error out when the MAX VALUE set conflicts with the integer type configure
2) let the MAX VALUE override what the default max is for the integer type.

Since my change deals with int64 (which is the largest encompassing integer type), option 2) seems reasonable to me. If/when we want to implement support for int2/int4
</blockquote></details>

Your comment looks out of date, looking at the state of the code below. In general, we would want to match Postgres as closely as possible. Postgres is your PM ;) 
___
*[pkg/sql/sequence.go, line 249 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/57339#-MOCIjrd7IWZYrBSYLDQ:-MOCIjrd7IWZYrBSYLDR:b-17uu5o) ([raw file](https://github.com/cockroachdb/cockroach/blob/84b98616cfe83ada143c0ab46fc291e3e753a900/pkg/sql/sequence.go#L249)):*
> ```Go
> 
> 	// All other defaults are dependent on the value of increment,
> 	// i.e. whether the sequence is ascending or descending.
> ```

This comment could benefit from an update now. 
___
*[pkg/sql/sequence.go, line 267 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/57339#-MOCJwu45f7UAdrZdyOX:-MOCJwu45f7UAdrZdyOY:b-ird0q0) ([raw file](https://github.com/cockroachdb/cockroach/blob/84b98616cfe83ada143c0ab46fc291e3e753a900/pkg/sql/sequence.go#L267)):*
> ```Go
> 	if setDefaults {
> 
> 		// TODO: Move this to a setDefaults helper function
> ```

Sounds like a good idea, are you planning on doing it as part of this PR? If not, I'd recommend adding someones name to the TODO. Maybe @adityamaru, considering you'll be leaving?
___
*[pkg/sql/sequence.go, line 278 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/57339#-MOCJ_tEC9nS8la0JzEl:-MOCJ_tEC9nS8la0JzEm:bjygcu8) ([raw file](https://github.com/cockroachdb/cockroach/blob/84b98616cfe83ada143c0ab46fc291e3e753a900/pkg/sql/sequence.go#L278)):*
> ```Go
> 			case types.Int2:
> 				opts.MaxValue = math.MaxInt16
> 				upperIntBound = math.MaxInt16
> ```

Should this be in the `setDefaults` if condition? Seems like you're using these bounds to check min/max value assigning later on. This would break the `ALTER SEQUENCE ... MAX VALUE ... ` case, if the max value being set is out of bounds, right?
___
*[pkg/sql/sequence_test.go, line 412 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/57339#-MOCPbHG42S5qQ-iAwGa:-MOCPbHG42S5qQ-iAwGb:b-mn27qa) ([raw file](https://github.com/cockroachdb/cockroach/blob/84b98616cfe83ada143c0ab46fc291e3e753a900/pkg/sql/sequence_test.go#L412)):*
> ```Go
> }
> 
> func TestCreateSequenceAsOption(t *testing.T) {
> ```

There's already a SHOW CREATE SEQUENCE test that you've added to in `show_test.go`. I'm wondering if all these test cases can just move there instead? As for the errors you're checking here, those should probably make their way to logictests. 
___
*[pkg/sql/show_test.go, line 475 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/57339#-MOCNfrBAk2PcDRh8KW6:-MOCNfrC8Y9a4m8wZepU:bilmfqh) ([raw file](https://github.com/cockroachdb/cockroach/blob/84b98616cfe83ada143c0ab46fc291e3e753a900/pkg/sql/show_test.go#L475)):*
> ```Go
> 		{
> 			`CREATE SEQUENCE %s AS bigint`,
> 			`CREATE SEQUENCE public.%s AS int MINVALUE 1 MAXVALUE 9223372036854775807 INCREMENT 1 START 1`,
> ```

This breaks round trip ability of SHOW CREATE SEQUENCE when the default int size is set to INT4 -- see my comments below. 
___
*[pkg/sql/catalog/descpb/structured.proto, line 962 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/57339#-MOCEZqv9dOuhGInXAFN:-MOCEZqv9dOuhGInXAFO:b-jt88md) ([raw file](https://github.com/cockroachdb/cockroach/blob/84b98616cfe83ada143c0ab46fc291e3e753a900/pkg/sql/catalog/descpb/structured.proto#L962)):*
> ```protobuf
>     optional SequenceOwner sequence_owner = 6 [(gogoproto.nullable) = false];
>     
>     optional string as_integer_type = 7 [(gogoproto.nullable) = false];
> ```

nit: Could you add a comment here. 
___
*[pkg/sql/parser/parse_test.go, line 1869 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/57339#-MOCLBHl3lTDLvyyAWft:-MOCLBHl3lTDLvyyAWfu:bku89ct) ([raw file](https://github.com/cockroachdb/cockroach/blob/84b98616cfe83ada143c0ab46fc291e3e753a900/pkg/sql/parser/parse_test.go#L1869)):*
> ```Go
> 		// `int` and `integer` parse to nakedIntTyoe which by default is Int/int64.
> 		{`CREATE SEQUENCE a AS integer`, `CREATE SEQUENCE a AS int`},
> 		{`CREATE SEQUENCE a AS bigint`, `CREATE SEQUENCE a AS int`},
> ```

This breaks round trip ability of SHOW CREATE TABLE?. Consider, `CREATE SEQUENCE AS BIGINT` which resolves to `CREATE SEQUENCE AS INT`, but INT can resolve to INT4 based on the cluster/session setting. 
___
*[pkg/sql/parser/sql.y, line 6489 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57339#-MNdHcFZ0Ti_Td45T-3i:-MOCF4yr465gJiPwc29T:bk59v2m) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9de7d968bc8a22b6e0899b66b8a3694ce508e15/pkg/sql/parser/sql.y#L6489)):*
<details><summary><i>Previously, mokaixu (Monica Xu) wrote…</i></summary><blockquote>

dare i say i'm creating ARUL?!?!??!?!?!?
</blockquote></details>

good one hehe :joy: 
___
*[pkg/sql/parser/sql.y, line 202 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/57339#-MOCFAFC69gSHfYbnh-h:-MOCFAFC69gSHfYbnh-i:bofwcq9) ([raw file](https://github.com/cockroachdb/cockroach/blob/84b98616cfe83ada143c0ab46fc291e3e753a900/pkg/sql/parser/sql.y#L202)):*
> ```Yacc
> 
> 
> func (u *sqlSymUnion) colType2() *types.T {
> ```

This looks unused. 
___
*[pkg/sql/parser/sql.y, line 6533 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/57339#-MOCRX5Q1qri80SZWwvQ:-MOCRX5Q1qri80SZWwvR:b-bc4yvi) ([raw file](https://github.com/cockroachdb/cockroach/blob/84b98616cfe83ada143c0ab46fc291e3e753a900/pkg/sql/parser/sql.y#L6533)):*
> ```Yacc
> // %Text:
> // CREATE [TEMPORARY | TEMP] SEQUENCE <seqname>
> //   [AS <typename>]
> ```

Could you add this docs thing to ALTER SEQUENCE as well! 
___
*[pkg/sql/parser/sql.y, line 9487 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/57339#-MOCGkF24CxPwxIQWrT2:-MOCGkF24CxPwxIQWrT3:bf5cqla) ([raw file](https://github.com/cockroachdb/cockroach/blob/84b98616cfe83ada143c0ab46fc291e3e753a900/pkg/sql/parser/sql.y#L9487)):*
> ```Yacc
>   }
> ```

nit: extra line
___
*[pkg/sql/sem/tree/create.go, line 1464 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/57339#-MOCKM4E3RKCimY7Cyq4:-MOCKM4E3RKCimY7Cyq5:bas7n92) ([raw file](https://github.com/cockroachdb/cockroach/blob/84b98616cfe83ada143c0ab46fc291e3e753a900/pkg/sql/sem/tree/create.go#L1464)):*
> ```Go
> // SequenceOptAs represents possible values (integer types)
> // the AS option of a CREATE SEQUENCE statement can have.
> type SequenceOptAs int
> ```

Is there a meaningful use of this anymore? Looks like `AsIntegerType` below does exactly what this would've aimed to accomplish.


<!-- Sent from Reviewable.io -->

mokaixu reviewed (commented):
Some notes in the most recent change:

* The SQL parsing now will only permit: BIGINT, 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57339#-:-MNyTnYm8xQeY8RN_-p7:b-752yw7)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru and @arulajmani)

___
*[pkg/sql/sequence.go, line 279 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57339#-MNdI1Oq80ZtIvE-h8dT:-MOhrct0CJCwMFC4qHch:b85p5dy) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9de7d968bc8a22b6e0899b66b8a3694ce508e15/pkg/sql/sequence.go#L279)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Your comment looks out of date, looking at the state of the code below. In general, we would want to match Postgres as closely as possible. Postgres is your PM ;) 
</blockquote></details>

Looks like Postgres errors out (approach #1)

See below

```
postgres=# CREATE SEQUENCE bad AS smallint MAXVALUE 200000;
ERROR:  MAXVALUE (200000) is out of range for sequence data type smallint
```

I've implemented that! but my message looks like this: "MAXVALUE (%d) must be less than (%d) for type %s",
___
*[pkg/sql/sequence.go, line 249 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/57339#-MOCIjrd7IWZYrBSYLDQ:-MOhhYTS8cjJzPhwaH_C:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/84b98616cfe83ada143c0ab46fc291e3e753a900/pkg/sql/sequence.go#L249)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

This comment could benefit from an update now. 
</blockquote></details>

done
___
*[pkg/sql/sequence.go, line 267 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/57339#-MOCJwu45f7UAdrZdyOX:-MOhmzXl3o7B6NfcKqLK:b8va3m4) ([raw file](https://github.com/cockroachdb/cockroach/blob/84b98616cfe83ada143c0ab46fc291e3e753a900/pkg/sql/sequence.go#L267)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Sounds like a good idea, are you planning on doing it as part of this PR? If not, I'd recommend adding someones name to the TODO. Maybe @adityamaru, considering you'll be leaving?
</blockquote></details>

i just refactored it into a helper method. 
Currently, I pass variables by reference and modify their value in the setSequenceDefaultVals method.

Another approach is to return the values of the variables. Which one do you think has more sensible design?
___
*[pkg/sql/sequence.go, line 278 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/57339#-MOCJ_tEC9nS8la0JzEl:-MOhsebZ8_JqseHtimB_:b-l5bqa3) ([raw file](https://github.com/cockroachdb/cockroach/blob/84b98616cfe83ada143c0ab46fc291e3e753a900/pkg/sql/sequence.go#L278)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Should this be in the `setDefaults` if condition? Seems like you're using these bounds to check min/max value assigning later on. This would break the `ALTER SEQUENCE ... MAX VALUE ... ` case, if the max value being set is out of bounds, right?
</blockquote></details>

To clarify, setDefaults is used only for CREATE SEQUENCE? Otherwise, we are in an ALTER SEQUENCE?

If that is the case, then i will move this logic pertaining to integerType outside of the setDefaults if statement
___
*[pkg/sql/sequence_test.go, line 412 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/57339#-MOCPbHG42S5qQ-iAwGa:-MOhpfvB2KiQYRiIfWqJ:be74fu5) ([raw file](https://github.com/cockroachdb/cockroach/blob/84b98616cfe83ada143c0ab46fc291e3e753a900/pkg/sql/sequence_test.go#L412)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

There's already a SHOW CREATE SEQUENCE test that you've added to in `show_test.go`. I'm wondering if all these test cases can just move there instead? As for the errors you're checking here, those should probably make their way to logictests. 
</blockquote></details>

Will look into writing logic tests 
___
*[pkg/sql/show_test.go, line 475 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/57339#-MOCNfrBAk2PcDRh8KW6:-MOhnVtQ0EJH6zVBXWJc:bj0j6dm) ([raw file](https://github.com/cockroachdb/cockroach/blob/84b98616cfe83ada143c0ab46fc291e3e753a900/pkg/sql/show_test.go#L475)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

This breaks round trip ability of SHOW CREATE SEQUENCE when the default int size is set to INT4 -- see my comments below. 
</blockquote></details>

Same response as below: how do we account for round trip ability when cluster settings are outside our control? Do we just want to hardcode  

```
seq_opt_as_int:
  INT
  {
    $$.val = sqllex.(*lexer).nakedIntType
  }
```

I adopted this from the `numeric` SQL type which is the following:

```
// SQL numeric data types
numeric:
  INT
  {
    $$.val = sqllex.(*lexer).nakedIntType
  }
| INTEGER
  {
    $$.val = sqllex.(*lexer).nakedIntType
  }
| SMALLINT
  {
    $$.val = types.Int2
  }
| BIGINT
  {
    $$.val = types.Int
  }
```
___
*[pkg/sql/catalog/descpb/structured.proto, line 962 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/57339#-MOCEZqv9dOuhGInXAFN:-MOhntjUESVlHbbJUz4l:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/84b98616cfe83ada143c0ab46fc291e3e753a900/pkg/sql/catalog/descpb/structured.proto#L962)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nit: Could you add a comment here. 
</blockquote></details>

done
___
*[pkg/sql/parser/parse_test.go, line 1869 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/57339#-MOCLBHl3lTDLvyyAWft:-MOhp77_Bu9LLk5wiqIy:b-7en9h2) ([raw file](https://github.com/cockroachdb/cockroach/blob/84b98616cfe83ada143c0ab46fc291e3e753a900/pkg/sql/parser/parse_test.go#L1869)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

This breaks round trip ability of SHOW CREATE TABLE?. Consider, `CREATE SEQUENCE AS BIGINT` which resolves to `CREATE SEQUENCE AS INT`, but INT can resolve to INT4 based on the cluster/session setting. 
</blockquote></details>

I'm not sure how to fix this since Cluster/Session settings are the reason why this isn't round trippable (nakedIntType can be int4 or int8)
___
*[pkg/sql/parser/sql.y, line 6489 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57339#-MNdHcFZ0Ti_Td45T-3i:-MOhoeeLBxUbhAYup7I0:b44tze) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9de7d968bc8a22b6e0899b66b8a3694ce508e15/pkg/sql/parser/sql.y#L6489)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

good one hehe :joy: 
</blockquote></details>

🦃
___
*[pkg/sql/parser/sql.y, line 202 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/57339#-MOCFAFC69gSHfYbnh-h:-MOhnxYQ9FY4JV_-XdJQ:b-1nhnlr) ([raw file](https://github.com/cockroachdb/cockroach/blob/84b98616cfe83ada143c0ab46fc291e3e753a900/pkg/sql/parser/sql.y#L202)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

This looks unused. 
</blockquote></details>

oops. deleted 
___
*[pkg/sql/parser/sql.y, line 6533 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/57339#-MOCRX5Q1qri80SZWwvQ:-MOhocrxDJqLWibBfnFE:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/84b98616cfe83ada143c0ab46fc291e3e753a900/pkg/sql/parser/sql.y#L6533)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Could you add this docs thing to ALTER SEQUENCE as well! 
</blockquote></details>

done
___
*[pkg/sql/parser/sql.y, line 9487 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/57339#-MOCGkF24CxPwxIQWrT2:-MOhoti44ixOcKhtvGKW:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/84b98616cfe83ada143c0ab46fc291e3e753a900/pkg/sql/parser/sql.y#L9487)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nit: extra line
</blockquote></details>

done
___
*[pkg/sql/sem/tree/create.go, line 1464 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/57339#-MOCKM4E3RKCimY7Cyq4:-MOhp08E3S3YOaEoM_Vr:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/84b98616cfe83ada143c0ab46fc291e3e753a900/pkg/sql/sem/tree/create.go#L1464)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Is there a meaningful use of this anymore? Looks like `AsIntegerType` below does exactly what this would've aimed to accomplish.
</blockquote></details>

done


<!-- Sent from Reviewable.io -->

arulajmani reviewed (changes requested):
@mokaixu and I talked offline about this and how we can make `SHOW CREATE SEQUENCE` round-tripable. The tldr here is that BIGINT would resolve to INT8 instead of INT when formatting so that the statement round trips independent of the int size cluster setting. Currently. INT8 isn't accepted as a valid datatype value for `CREATE SEQUENCE AS`, but we confirmed that PG accepts it even though it doesn't explicitly state this in its docs. As part of making this roundtripable, we'll also accept INT8, INT4, etc. as possible values in `CREATE SEQUENCE AS`, to be in line w PG.

I can take another pass after we have the new approach implemented and more logictests -- thanks for iterating on this! 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57339#-:-MOrglBLDtessPlq8vLI:b-ewd2ej)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru and @mokaixu)

___
*[pkg/sql/sequence.go, line 279 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57339#-MNdI1Oq80ZtIvE-h8dT:-MOsE9S1DU7jM0tVsDQO:b-x9sdxa) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9de7d968bc8a22b6e0899b66b8a3694ce508e15/pkg/sql/sequence.go#L279)):*
<details><summary><i>Previously, mokaixu (Monica Xu) wrote…</i></summary><blockquote>

Looks like Postgres errors out (approach #1)

See below

```
postgres=# CREATE SEQUENCE bad AS smallint MAXVALUE 200000;
ERROR:  MAXVALUE (200000) is out of range for sequence data type smallint
```

I've implemented that! but my message looks like this: "MAXVALUE (%d) must be less than (%d) for type %s",
</blockquote></details>

minor nit would be to match the pg error message as closely as possible, but otherwise this looks good. 
___
*[pkg/sql/sequence.go, line 278 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/57339#-MOCJ_tEC9nS8la0JzEl:-MOrlCqLAkJdsHT3KBLb:b-indwo1) ([raw file](https://github.com/cockroachdb/cockroach/blob/84b98616cfe83ada143c0ab46fc291e3e753a900/pkg/sql/sequence.go#L278)):*
<details><summary><i>Previously, mokaixu (Monica Xu) wrote…</i></summary><blockquote>

To clarify, setDefaults is used only for CREATE SEQUENCE? Otherwise, we are in an ALTER SEQUENCE?

If that is the case, then i will move this logic pertaining to integerType outside of the setDefaults if statement
</blockquote></details>

Yeah, exactly! `ALTER SEQUENCE` passes false to `setDefaults` above. And yes, what you're saying is exactly what I was alluding to, which is why I am encouraging you to write logictests for both create sequence and alter sequence with your new changes! 
___
*[pkg/sql/sequence_test.go, line 412 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/57339#-MOCPbHG42S5qQ-iAwGa:-MOrlhO--tR8x-iLWVTX:b-ylsncu) ([raw file](https://github.com/cockroachdb/cockroach/blob/84b98616cfe83ada143c0ab46fc291e3e753a900/pkg/sql/sequence_test.go#L412)):*
<details><summary><i>Previously, mokaixu (Monica Xu) wrote…</i></summary><blockquote>

Will look into writing logic tests 
</blockquote></details>

Okay, once you do, you might be able to come back here and clean this stuff up completely. 
___
*[pkg/sql/show_test.go, line 475 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/57339#-MOCNfrBAk2PcDRh8KW6:-MOsFUjLELPAR93Mfw7K:b-kpmhzv) ([raw file](https://github.com/cockroachdb/cockroach/blob/84b98616cfe83ada143c0ab46fc291e3e753a900/pkg/sql/show_test.go#L475)):*
<details><summary><i>Previously, mokaixu (Monica Xu) wrote…</i></summary><blockquote>

Same response as below: how do we account for round trip ability when cluster settings are outside our control? Do we just want to hardcode  

```
seq_opt_as_int:
  INT
  {
    $$.val = sqllex.(*lexer).nakedIntType
  }
```

I adopted this from the `numeric` SQL type which is the following:

```
// SQL numeric data types
numeric:
  INT
  {
    $$.val = sqllex.(*lexer).nakedIntType
  }
| INTEGER
  {
    $$.val = sqllex.(*lexer).nakedIntType
  }
| SMALLINT
  {
    $$.val = types.Int2
  }
| BIGINT
  {
    $$.val = types.Int
  }
```
</blockquote></details>

No longer applicable after our discussion offline -- see above.
___
*[pkg/sql/parser/parse_test.go, line 1869 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/57339#-MOCLBHl3lTDLvyyAWft:-MOsEL8b-f8y_lBGlmZ0:bawxzzh) ([raw file](https://github.com/cockroachdb/cockroach/blob/84b98616cfe83ada143c0ab46fc291e3e753a900/pkg/sql/parser/parse_test.go#L1869)):*
<details><summary><i>Previously, mokaixu (Monica Xu) wrote…</i></summary><blockquote>

I'm not sure how to fix this since Cluster/Session settings are the reason why this isn't round trippable (nakedIntType can be int4 or int8)
</blockquote></details>

not applicable anymore, see summary of discussion above.


<!-- Sent from Reviewable.io -->

mokaixu reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57339)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru and @arulajmani)

___
*[pkg/sql/sequence.go, line 279 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57339#-MNdI1Oq80ZtIvE-h8dT:-MP6grC15_xoYlYuv1WQ:b-4cdk05) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9de7d968bc8a22b6e0899b66b8a3694ce508e15/pkg/sql/sequence.go#L279)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

minor nit would be to match the pg error message as closely as possible, but otherwise this looks good. 
</blockquote></details>

I will address this as soon as CI passes!
___
*[pkg/sql/sequence.go, line 278 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/57339#-MOCJ_tEC9nS8la0JzEl:-MP6flhy26a89YLpqYXS:bv84tih) ([raw file](https://github.com/cockroachdb/cockroach/blob/84b98616cfe83ada143c0ab46fc291e3e753a900/pkg/sql/sequence.go#L278)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Yeah, exactly! `ALTER SEQUENCE` passes false to `setDefaults` above. And yes, what you're saying is exactly what I was alluding to, which is why I am encouraging you to write logictests for both create sequence and alter sequence with your new changes! 
</blockquote></details>

doubly done!
___
*[pkg/sql/sequence_test.go, line 412 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/57339#-MOCPbHG42S5qQ-iAwGa:-MP6fvg01T4DxcIXroDu:bh3dua1) ([raw file](https://github.com/cockroachdb/cockroach/blob/84b98616cfe83ada143c0ab46fc291e3e753a900/pkg/sql/sequence_test.go#L412)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Okay, once you do, you might be able to come back here and clean this stuff up completely. 
</blockquote></details>

I've deleted redundant tests in `sequence_test.go` compared to `show_test.go` . `sequence_test.go` now primarily tests basic sequence functions and possible errors that can come up in the business logic.
___
*[pkg/sql/show_test.go, line 475 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/57339#-MOCNfrBAk2PcDRh8KW6:-MP6gkyaDgKa73w0J6TY:b2nhysk) ([raw file](https://github.com/cockroachdb/cockroach/blob/84b98616cfe83ada143c0ab46fc291e3e753a900/pkg/sql/show_test.go#L475)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

No longer applicable after our discussion offline -- see above.
</blockquote></details>

ack
___
*[pkg/sql/parser/parse_test.go, line 1869 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/57339#-MOCLBHl3lTDLvyyAWft:-MP6gmMO42AoBswWQWci:b2nhysk) ([raw file](https://github.com/cockroachdb/cockroach/blob/84b98616cfe83ada143c0ab46fc291e3e753a900/pkg/sql/parser/parse_test.go#L1869)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

not applicable anymore, see summary of discussion above.
</blockquote></details>

ack


<!-- Sent from Reviewable.io -->

mokaixu reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57339)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru and @arulajmani)

___
*[pkg/sql/sequence.go, line 279 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57339#-MNdI1Oq80ZtIvE-h8dT:-MPANE_614oQJ3MmrRVu:bxttbuo) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9de7d968bc8a22b6e0899b66b8a3694ce508e15/pkg/sql/sequence.go#L279)):*
<details><summary><i>Previously, mokaixu (Monica Xu) wrote…</i></summary><blockquote>

I will address this as soon as CI passes!
</blockquote></details>

Hmm, I thought about it a little more and I think it's more informative to give the exact upper and lower bound. Otherwise, users would have to look up the upper and lower bounds for int2/int4/int8. What do you think?


<!-- Sent from Reviewable.io -->

adityamaru commented:
@arulajmani I'll be taking this over. I think Monica addressed your last round of comments so this should be RFAL, thanks!

arulajmani reviewed (changes requested):
Yo! Sorry for taking so long to get back on reviewing this thing -- OOO and what not.

Took another look, and most of my comments now are around testing. Specifically, I'd feel much better about this change once `ALTER SEQUENCE` is tested more thoroughly as well. I've left some suggestions in comments about ways to do this. Other than that, I think this is close. 

> Reviewed 2 of 13 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57339#-:-MQnXVSQ5y0iHeE-pnbc:b-wg1oqi)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @arulajmani, and @mokaixu)

___
*[pkg/sql/sequence_test.go, line 411 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/57339#-MQnS0i5DdmqMuuCcGam:-MQnS0i5DdmqMuuCcGan:b-l51qy3) ([raw file](https://github.com/cockroachdb/cockroach/blob/361311bd1d4920ed8641d139df30005cf2272bc7/pkg/sql/sequence_test.go#L411)):*
> ```Go
> 	require.NoError(t, err)
> }
> ```

@adityamaru can these tests not move to logic tests? A lot of what seems to be tested here is also tested in the logic tests, but I'm not sure if every case is. I think it's worth removing this entire function here and moving all the interesting tests to logic tests. 
___
*[pkg/sql/logictest/testdata/logic_test/sequences, line 1277 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/57339#-MQnTtVT-5vsMDT7gTFd:-MQnTtVT-5vsMDT7gTFe:b-3lsqxz) ([raw file](https://github.com/cockroachdb/cockroach/blob/361311bd1d4920ed8641d139df30005cf2272bc7/pkg/sql/logictest/testdata/logic_test/sequences#L1277)):*
> ```
> statement ok
> CREATE SEQUENCE db2.seq2 OWNED BY db1.t.a
> ```

We don't seem to be doing anything interesting with these that isn't covered by parse tests, right?
___
*[pkg/sql/logictest/testdata/logic_test/sequences, line 1331 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/57339#-MQnVpgE58toD-efSGNl:-MQnVpgF7qXIDlL4UaYq:bf1roo) ([raw file](https://github.com/cockroachdb/cockroach/blob/361311bd1d4920ed8641d139df30005cf2272bc7/pkg/sql/logictest/testdata/logic_test/sequences#L1331)):*
> ```
> 
> statement ok
> ALTER SEQUENCE seqas_9 AS smallint
> ```

Would be nice to get more `ALTER` tests (maybe in the `alter_sequence` file?). In particular, I'm thinking of cases where altering to a larger/smaller Integer type would cause errors because of previously set MAXVALUE/MINVALUE/START etc.

Another interesting test idea would be to try `ALTER`-ing a sequence to a smaller Integer type when the curval is greater than the max permitted.  


<!-- Sent from Reviewable.io -->

otan commented:
i gotchu @mokaixu ! miss the memes.

i've updated the PR with arul's suggestions, the ALTER SEQUENCE bugs are fixed in the second commit. since arul's changed teams, maybe @rafiss can take a look.

rafiss reviewed (changes requested):
awesome! thanks for bringing this back

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57339#-:-MogUbDNDqhsMyTSLcBb:b-dekom)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @arulajmani, @mokaixu, and @otan)

___
*[pkg/sql/sequence_test.go, line 861 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/57339#-MogUR4F5HO7kgXc8OBj:-MogUR4F5HO7kgXc8OBk:b-5ix57g) ([raw file](https://github.com/cockroachdb/cockroach/blob/4a089b9c3b320b8dc03bac12521ab1b9085a4b31/pkg/sql/sequence_test.go#L861)):*
> ```Go
> 			verifyQuery: `SHOW CREATE SEQUENCE a_seq`,
> 			expected: [][]string{{
> 				"a_seq", `CREATE SEQUENCE public.a_seq AS INT8 MINVALUE 0 MAXVALUE 234567 INCREMENT 1 START 2`,
> ```

not sure -- from looking at most of these, it seems nay would work as logic tests. are there any that can't?
___
*[pkg/sql/catalog/descpb/structured.proto, line 1100 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/57339#-MogTU0cC8SB4F-ADu-Y:-MogTU0cC8SB4F-ADu-Z:bth3bk1) ([raw file](https://github.com/cockroachdb/cockroach/blob/4a089b9c3b320b8dc03bac12521ab1b9085a4b31/pkg/sql/catalog/descpb/structured.proto#L1100)):*
> ```protobuf
>     // AS option value for CREATE SEQUENCE, which specifies the default
>     // min and max values a sequence can take on.
>     optional string as_integer_type = 7 [(gogoproto.nullable) = false];
> ```

hm this part seems backward incompatible. shouldn't `as_integer_type` be number 8?


<!-- Sent from Reviewable.io -->

otan committed:
sql: fix ALTER SEQUENCE ... AS ... defaults

Previously, ALTER SEQUENCE ... AS ... can change MINVALUE and MAXVALUE to
the new type value. This is now preserved and errors if values are out
of range.

Release note: None

otan reviewed (commented):
fixed up some more

> Reviewed 1 of 13 files at r4, 4 of 17 files at r5, 1 of 1 files at r6, 3 of 6 files at r8.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57339#-:-MohCiLH-nEHXN8X45ZI:b-n8nihf)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @arulajmani, @mokaixu, @otan, and @rafiss)

___
*[pkg/sql/sequence_test.go, line 412 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/57339#-MOCPbHG42S5qQ-iAwGa:-Moa2xXD56vnu1vGytsE:b-7ons9j) ([raw file](https://github.com/cockroachdb/cockroach/blob/84b98616cfe83ada143c0ab46fc291e3e753a900/pkg/sql/sequence_test.go#L412)):*
<details><summary><i>Previously, mokaixu (Monica Xu) wrote…</i></summary><blockquote>

I've deleted redundant tests in `sequence_test.go` compared to `show_test.go` . `sequence_test.go` now primarily tests basic sequence functions and possible errors that can come up in the business logic.
</blockquote></details>

the logic tests look pretty exhaustive so i removed this.
___
*[pkg/sql/sequence_test.go, line 411 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/57339#-MQnS0i5DdmqMuuCcGam:-Moa2t4B8-V-Cby3sHme:bvsr5nk) ([raw file](https://github.com/cockroachdb/cockroach/blob/361311bd1d4920ed8641d139df30005cf2272bc7/pkg/sql/sequence_test.go#L411)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

@adityamaru can these tests not move to logic tests? A lot of what seems to be tested here is also tested in the logic tests, but I'm not sure if every case is. I think it's worth removing this entire function here and moving all the interesting tests to logic tests. 
</blockquote></details>

I've done this.
___
*[pkg/sql/catalog/descpb/structured.proto, line 1100 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/57339#-MogTU0cC8SB4F-ADu-Y:-Moh1RmR72RDTQOcctaP:b-b8drlj) ([raw file](https://github.com/cockroachdb/cockroach/blob/4a089b9c3b320b8dc03bac12521ab1b9085a4b31/pkg/sql/catalog/descpb/structured.proto#L1100)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

hm this part seems backward incompatible. shouldn't `as_integer_type` be number 8?
</blockquote></details>

i fixed this in a later revision - can confirm version-upgrade tests work!
bad rebase :(


<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
lgtm with one question

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57339#-:-MojygqA8EEnIqsZQeN4:b-lqqpt9)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @arulajmani, @mokaixu, @otan, and @rafiss)

___
*[pkg/sql/sequence.go, line 396 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/57339#-MojxM0IEuGFX3t5ppbK:-MojxM0JBatuNrT7yytY:b-3euqw5) ([raw file](https://github.com/cockroachdb/cockroach/blob/57a01a57711ec826355d9a4063293d00b5dc20a1/pkg/sql/sequence.go#L396)):*
> ```Go
> 
> 	if isAscending {
> 		minValue = 1
> ```

why does isAscending change the minValue? is that just a weird quirk of how PG works too?


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
bors r=rafiss

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57339#-:-MokUP15Cyn-zC_9kHg7:bx70i2y)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @arulajmani, @mokaixu, @otan, and @rafiss)

___
*[pkg/sql/sequence.go, line 396 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/57339#-MojxM0IEuGFX3t5ppbK:-MokUIceEYEbON9Jrg7z:bp6ec9q) ([raw file](https://github.com/cockroachdb/cockroach/blob/57a01a57711ec826355d9a4063293d00b5dc20a1/pkg/sql/sequence.go#L396)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

why does isAscending change the minValue? is that just a weird quirk of how PG works too?
</blockquote></details>

yep. 1 to max_int and -1 to -min_int is the increment styles.


<!-- Sent from Reviewable.io -->

otan commented:
bors r=rafiss
